Sources/OpenCombine/Publishers/Publishers.CompactMap.swift

```diff
@@ -0,0 +1,217 @@
+        return .init(upstream: upstream,
+                     transform: { try self.transform($0).flatMap(transform) })
+    }
+}
+
+private class _CompactMap<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      Subscription
+{
+    typealias Input = Upstream.Output
+    typealias Failure = Upstream.Failure
+    typealias Transform = (Input) -> Result<Downstream.Input?, Downstream.Failure>
+
+    fileprivate var _transform: Transform?
+
+    private var _unsatisfiedDemand: Subscribers.Demand = .none
+
+    var _isCompleted: Bool {
+        return _transform == nil
+    }
+
+    init(downstream: Downstream, transform: @escaping Transform) {
+        _transform = transform
+        super.init(downstream: downstream)
+    }
+
+    func receive(subscription: Subscription) {
+        upstreamSubscription = subscription
+        downstream.receive(subscription: self)
+    }
+
+    func receive(_ input: Input) -> Subscribers.Demand {
+        switch _transform?(input) {
+        case .success(let output?)?:
+            _unsatisfiedDemand -= 1
+            let newDemand = downstream.receive(output)
+            _unsatisfiedDemand += newDemand
+            return newDemand
+        case .success(nil)?:
+            return .max(1)
+        case .failure(let error)?:
+            downstream.receive(completion: .failure(error))
+            _transform = nil
+            return .none
+        case nil:
+            return .none
+        }
+    }
+
+    func request(_ demand: Subscribers.Demand) {
+        guard !_isCompleted else { return }
+        _unsatisfiedDemand += demand
+        upstreamSubscription?.request(demand)
+    }
+
+    override func cancel() {
+        _transform = nil
+        upstreamSubscription?.cancel()
+        upstreamSubscription = nil
+    }
+}
+
+extension Publishers.CompactMap {
+    private final class Inner<Downstream: Subscriber>
+        : _CompactMap<Upstream, Downstream>,
+          Subscriber,
+          CustomStringConvertible
+        where Downstream.Failure == Upstream.Failure
+    {
+        func receive(completion: Subscribers.Completion<Upstream.Failure>) {
+            if !_isCompleted {
+                _transform = nil
+                downstream.receive(completion: completion)
+            }
+        }
+
+        var description: String { return "CompactMap" }
+    }
+}
+
+extension Publishers.TryCompactMap {
+    private final class Inner<Downstream: Subscriber>
+        : _CompactMap<Upstream, Downstream>,
+          Subscriber,
+          CustomStringConvertible
+        where Downstream.Failure == Error
+    {
+        func receive(completion: Subscribers.Completion<Upstream.Failure>) {
+            if !_isCompleted {
+                _transform = nil
+                downstream.receive(completion: completion.eraseError())
+            }
+        }
+
+        var description: String { return "TryCompactMap" }
```

Comment:
@broadwaylamb Can you put the properties at the top above the functions in the bodies of these helper classes?